tensorflow/core/util/ctc/ctc_decoder.h

```diff
@@ -69,25 +70,27 @@ class CTCDecoder {
 
 // CTCGreedyDecoder is an implementation of the simple best path decoding
 // algorithm, selecting at each timestep the most likely class at each timestep.
-class CTCGreedyDecoder : public CTCDecoder {
+template<class T>
+class CTCGreedyDecoder : public CTCDecoder<T> {
  public:
+  typedef CTCDecoder<T> Decoder;
   CTCGreedyDecoder(int num_classes, int batch_size, bool merge_repeated)
-      : CTCDecoder(num_classes, batch_size, merge_repeated) {}
+      : CTCDecoder<T>(num_classes, batch_size, merge_repeated) {}
 
-  Status Decode(const CTCDecoder::SequenceLength& seq_len,
-                const std::vector<CTCDecoder::Input>& input,
-                std::vector<CTCDecoder::Output>* output,
-                CTCDecoder::ScoreOutput* scores) override {
-    if (output->empty() || (*output)[0].size() < batch_size_) {
+  Status Decode(const typename CTCDecoder<T>::SequenceLength& seq_len,
+                const std::vector<typename CTCDecoder<T>::Input>& input,
+                std::vector<typename CTCDecoder<T>::Output>* output,
+                typename CTCDecoder<T>::ScoreOutput* scores) override {
+    if (output->empty() || (*output)[0].size() < Decoder::batch_size_) {
       return errors::InvalidArgument(
           "output needs to be of size at least (1, batch_size).");
     }
-    if (scores->rows() < batch_size_ || scores->cols() == 0) {
+    if (scores->rows() < Decoder::batch_size_ || scores->cols() == 0) {
       return errors::InvalidArgument(
           "scores needs to be of size at least (batch_size, 1).");
     }
     // For each batch entry, identify the transitions
-    for (int b = 0; b < batch_size_; ++b) {
+    for (int b = 0; b < Decoder::batch_size_; ++b) {
```

Comment:
ditto; undo

Reply:
same as in https://github.com/tensorflow/tensorflow/pull/31164#discussion_r312368512